pkg/k8s/shared_resources.go

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package k8s
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	cilium_api_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	// SharedResourceCell provides a set of shared handles to Kubernetes resources used throughout the
+	// Cilium agent. Each of the resources share a client-go informer and backing store so we only
+	// have one watch API call for each resource kind and that we maintain only one copy of each object.
+	//
+	// If you need a filtered view to a resource (e.g. objects in certain namespace) and it already
+	// has a more general resource, then use stream.Filter() to filter the event stream rather than
+	// creating a new resource.
+	//
+	// See pkg/k8s/resource/resource.go for documentation on the Resource[T] type.
+	SharedResourcesCell = cell.Module(
+		"k8s-shared-resources",
+		"Shared Kubernetes resources",
+
+		cell.Provide(
+			serviceResource,
+			localNodeResource,
+			namespaceResource,
+			lbIPPoolsResource,
+		),
+	)
+)
+
+type SharedResources struct {
```

Comment:
I have to say that this list feels somewhat arbitrary at the moment. What about `v2.CiliumNode` for example which is used in stead of `corev1.Node` depending on the current IPAM mode.

Also why is `LBIPPools` here? Why would we need to share that resource so broadly?

I would say we take a stance, either all resource definitions (that are actually reusable) should be defined here or only the most generic ones. Perhaps we can clarify intended usage with a comment on SharedResources

Reply:
Well, it's not arbitrary, it's just incomplete list. My thinking was that we'd have all of them in one place so it's easy to see to what's all the things we're potentially subscribing to and what all their options are. E.g. it's a bit trying to be a "SharedInformerFactory" in client-go.

There might be cases in the future where we actually don't want to share a resource, e.g. I could imagine that we want `Service` and `Endpoint` to first be coalesced into ServiceCache and that would expose access to them, but for now it feels like it's safer bet to keep all of them in one place.